### PR TITLE
Adding and removing users to/from groups

### PR DESCRIPTION
**Pull request by bp323**
It would be really useful if users could be added to groups in the UI (both by a global admin and an application admin).

Currently, groups don't have a name, but since the idea is that a user would "manage" a part, we can simply display the course, subject and part tree and show a little checkbox next to each part.

Something like:
```
Advanced computer science
    ☐ MPhil
American Literature
    ☐ Mphil
Asian and middle eastern studies
   Chinese studies
       ☐ Part IA
       ☐ Part IB
       ☑ Part II
   Japanese studies
       ☐ Part IA
       ☐ Part IB
       ☑ Part II
...
```

### CODE REVIEW

**Comment by bp323** [comment]
Converted into a PR for visibility

**Comment by bp323** [comment]
Depends on https://github.com/fronteerio/grasshopper/tree/orgunit-includepermissions

File: apps/admin/ui/css/admin.css [hunk 1]
```diff
@@ -23,103 +23,15 @@ main h2 {
...
+    margin: 20px;
 }
 
-.gh-administration-container .panel-default .panel-heading .gh-striped-container-search button {
-    height: 34px;
-    padding: 6px 12px;
-}
+.gh-administration-container .gh-configuration-form {
+    margin-bottom: 15px !important;
+}
```
Comment (bp323) [inline apps/admin/ui/css/admin.css]: Will fix this new line

**Comment by bp323** [comment]
Reassigning to @Coenego for a round of review. Note that the backend will change and this PR will have to follow.

**Comment by bp323** [comment]
Depends on https://github.com/fronteerio/grasshopper/pull/101

File: shared/gh/js/controllers/gh.api.user.js [hunk 2]
```diff
@@ -69,7 +64,7 @@ define(['exports'], function(exports) {
                 return callback({'code': jqXHR.status, 'msg': jqXHR.responseText});
             }
         });
-        */
+
```
Comment (Coenego) [inline shared/gh/js/controllers/gh.api.user.js]: Empty line can probably go

File: shared/gh/js/controllers/gh.api.orgunit.js [hunk 3]
```diff
@@ -357,14 +357,14 @@ define(['exports'], function(exports) {
...
-     * @param  {Number}      [parentId]              The ID of the parent to retrieve the organisational units for
-     * @param  {String[]}    [type]                  The organisational unit type(s) to filter the organisational unit by
-     * @param  {Function}    callback                Standard callback function
-     * @param  {Object}      callback.err            Error object containing the error code and error message
-     * @param  {Object}      callback.response       Object representing the organisational units in an app
+     * @param  {Number}             [appId]                 The ID of the app to get the organisational units for
+     * @param  {Boolean}            [includeSeries]         Whether to include the event series associated to the oranisational units
+     * @param  {Boolean|Number}     [includePermissions]    Whether to include if the current user can manage the organisational units/series and whether an organisational unit is locked. If a Number instead of a Boolean is provided the permissions for the specified user will be returned. Defaults to `false`
```
Comment (Coenego) [inline shared/gh/js/controllers/gh.api.orgunit.js]: Only 4 spaces allowed after the type declaration

File: shared/gh/css/gh.skin.css [hunk 4]
```diff
@@ -355,6 +355,25 @@ ul.nav-tabs li.active {
...
+.panel-group:last-child .panel-default {
+    border-color: #DDD;
+}
+
+/* Striped rows */
+
+.panel-default .panel-body .gh-striped-row:nth-child(odd) {
+    background-color: #F9F9F9;
+}
+
```
Comment (Coenego) [inline shared/gh/css/gh.skin.css]: Extra empty line needed

File: apps/timetable/admin/app/js/index.js [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+            gh.api.groupsAPI.updateGroupMembers(change.id, changeObj, function(err) {
+                if (err) {
+                    return gh.utils.notification('Could not update user access', constants.messaging.default.error, 'error');
+                }
+
+                done++;
+                if (done === todo) {
+                    callback();
+                } else {
+                    updateGroupMember(changeSet[done], callback);
```
Comment (Coenego) [inline apps/timetable/admin/app/js/index.js]: We can probably git rid off the `else` by returning the `callback();`

File: apps/timetable/admin/app/js/index.js [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+        _.each($('[type="checkbox"]:not(:checked)', $form), function(chk) {
+            configValues[$(chk).attr('name')] = $(chk).is(':checked');
+        });
+
+        // Update the configuration
+        gh.api.configAPI.updateConfig($form.data('appid'), configValues, function(err) {
+            if (err) {
+                return gh.utils.notification('App configuration not updated', constants.messaging.default.error, 'error');
+            }
+            return gh.utils.notification('App configuration updated', null, 'success');
```
Comment (Coenego) [inline apps/timetable/admin/app/js/index.js]: unnecessary `return` here

File: apps/timetable/admin/app/js/index.js [hunk 7]
```diff
@@ -0,0 +1,623 @@
...
+        var displayName = $('.gh-app-displayname', $(this)).val();
+        var enabled = $('.gh-app-enabled', $(this)).is(':checked');
+        var host = $('.gh-app-host', $(this)).val();
+
+        // Update the app
+        gh.api.appAPI.updateApp(appId, displayName, enabled, host, function(err, data) {
+            if (err) {
+                return gh.utils.notification('Could not update the system app', constants.messaging.default.error, 'error');
+            }
+
```
Comment (Coenego) [inline apps/timetable/admin/app/js/index.js]: Empty line can probably go

File: shared/gh/js/controllers/gh.api.user.js [hunk 8]
```diff
@@ -69,7 +64,34 @@ define(['exports'], function(exports) {
...
+    };
+
+    /**
+     * Get the groups a user is a member of
+     *
+     * @param  {[type]}      userId               The ID of the user to get the group memberships for
+     * @param  {Function}    callback             Standard callback function
+     * @param  {Object}      callback.err         Error object containing the error code and error message
+     * @param  {Object}      callback.response    The groups the user is a member of
+     */
```
Comment (Coenego) [inline shared/gh/js/controllers/gh.api.user.js]: I've been using `@throws` as well lately. Shall we add it wherever we thrown errors for parameter validation?

File: tests/qunit/tests/js/api.user.js [hunk 9]
```diff
@@ -135,20 +135,78 @@ require(['gh.core', 'gh.api.tests'], function(gh, testAPI) {
...
+
+                                    QUnit.start();
+                                });
+                            });
+                        });
+                    });
+                });
+            });
+        });
+
```
Comment (Coenego) [inline tests/qunit/tests/js/api.user.js]: Empty line can probably go

**Comment by Coenego** [comment]
Back to @bp323 for follow-up after a first round of review.

File: shared/gh/js/controllers/gh.api.user.js (hunk 8 above)
Comment (bp323) [inline shared/gh/js/controllers/gh.api.user.js]: Agreed but let's do it in a separate PR

**Comment by bp323** [comment]
Looks like the backend for this has been merged, ready for review.

File: apps/timetable/admin/app/js/index.js [hunk 10]
```diff
@@ -0,0 +1,622 @@
...
+     */
+    var renderAppUserSearch = function() {
+        gh.utils.renderTemplate($('#gh-app-user-search-template'), null, $('#gh-main'));
+    };
+
+    /**
+     * Render the tripos structure indicating which groups the selected user is part of
+     *
+     * @param  {Object}    triposData    The app's tripos structure
+     * @private
```
Comment (Coenego) [inline apps/timetable/admin/app/js/index.js]: Missing parameter for `user` and `memberships`

File: apps/timetable/admin/app/js/index.js [hunk 11]
```diff
@@ -0,0 +1,622 @@
...
+    var renderApp = function() {
+        gh.utils.renderTemplate($('#gh-app-template'), {
+            'app': gh.data.me.app
+        }, $('#gh-main'));
+    };
+
+    /**
+     * Render the configuration form
+     *
+     * @private
```
Comment (Coenego) [inline apps/timetable/admin/app/js/index.js]: Missing parameter for `config`

File: apps/timetable/admin/app/js/index.js [hunk 12]
```diff
@@ -0,0 +1,622 @@
...
+
+
+    ///////////
+    // USERS //
+    ///////////
+
+    /**
+     * Get the user object from the user AutoSuggest field
+     *
+     * @return {User[]}    User object chosen in the user AutoSuggest field
```
Comment (Coenego) [inline apps/timetable/admin/app/js/index.js]: This should probably be `@return {User}` instead of `@return {User[]}`

File: apps/timetable/admin/app/js/index.js [hunk 13]
```diff
@@ -0,0 +1,622 @@
...
+            'id': parseInt($user.data('value').toString(), 10),
+            'displayName': $user.text().slice(1)
+        };
+    };
+
+    /**
+     * Submit the form with parts the user has access to
+     *
+     * @private
+     * @return {Boolean}    Return false to avoid the default form submit behaviour
```
Comment (Coenego) [inline apps/timetable/admin/app/js/index.js]: `@return` should go before `@private`

File: apps/timetable/apache/app.conf [hunk 14]
```diff
@@ -35,6 +35,21 @@
     # App files
     Alias /apps <%= apacheConf.timetable.apps %>
 
+   # App user management
```
Comment (Coenego) [inline apps/timetable/apache/app.conf]: Indention issue

File: shared/gh/js/controllers/gh.api.user.js (hunk 8 above)
Comment (Coenego) [inline shared/gh/js/controllers/gh.api.user.js]: Implemented in https://github.com/CUL-DigitalServices/grasshopper-ui/pull/376

**Comment by Coenego** [comment]
Looks good! Some minor comments left. Back to @bp323 for follow-up.

**Comment by bp323** [comment]
Over to @Coenego for review.
